Replace deprecated "NSErrorFailingURLStringKey" and "NSURLErrorFailingURLStringErrorKey" with "NSURLErrorFailingURLErrorKey"

jelee53 opened the pull request:
#### 4f50f9ee9eaac22f0e380d3ce464746fd6dc69e0
<pre>
Replace deprecated &quot;NSErrorFailingURLStringKey&quot; and &quot;NSURLErrorFailingURLStringErrorKey&quot; with &quot;NSURLErrorFailingURLErrorKey&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=297139">https://bugs.webkit.org/show_bug.cgi?id=297139</a>
<a href="https://rdar.apple.com/158015996">rdar://158015996</a>

Reviewed by Sihui Liu and Per Arne Vollan.

Remove usage of  &quot;NSErrorFailingURLStringKey&quot; and &quot;NSURLErrorFailingURLStringErrorKey&quot; in up-to-date platforms

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/network/cf/ResourceError.h:
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::createNSErrorFromResourceErrorBase):
(WebCore::ResourceError::hasMatchingFailingURLKeys const):
* Source/WebKit/Shared/Cocoa/WebErrorsCocoa.mm:
(WebKit::createNSError):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadInvalidURLRequest.mm:
(TestWebKitAPI::TEST(WebKit, LoadInvalidURLWithSpaceCharacter)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeBrowsing.mm:
(TEST(SafeBrowsing, WKWebViewGoBackIFrame)):
(TEST(SafeBrowsing, PostResponseIframe)):

Canonical link: <a href="https://commits.webkit.org/298727@main">https://commits.webkit.org/298727@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a013956b6ae5a6ffc2ecced450e20da061cbcf73

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116450 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36111 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26662 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122507 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67013 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118339 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36805 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44699 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88435 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119399 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29356 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104471 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68876 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28428 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22577 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66188 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108560 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98731 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22738 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125656 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114978 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43344 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32550 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97145 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43709 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100680 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96940 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42232 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20141 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39298 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18597 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43232 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48823 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143676 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42698 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37002 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46038 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44403 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jelee53 commented on Source/WebKitLegacy/mac/Misc/WebKitErrors.m:
```diff
@@ -74,7 +74,7 @@ - (instancetype)_webkit_initWithDomain:(NSString *)domain code:(int)code URL:(NS
     ALLOW_DEPRECATED_DECLARATIONS_BEGIN
     NSDictionary *userInfo = [NSDictionary dictionaryWithObjectsAndKeys:
         URL, NSURLErrorFailingURLErrorKey,
-        [URL absoluteString], @"NSErrorFailingURLStringKey",
```
TODO: This is Webkit Legacy Maybe we should leave this one as-is.

jelee53 commented on WebKitLibraries/SDKDBs/iphoneos18.5/Foundation.partial.sdkdb:
```diff
@@ -4214,7 +4214,7 @@
     {"access": "public", "kind": "function", "linkage": "exported", "name": "_NSEnumerateMapTable"},
     {"access": "public", "kind": "function", "linkage": "internal", "name": "_NSEqualRanges"},
-    {"access": "public", "introduced": "2", "kind": "variable", "linkage": "exported", "name": "_NSErrorFailingURLStringKey"},
+    {"access": "public", "introduced": "2", "kind": "variable", "linkage": "exported", "name": "_NSURLErrorFailingURLErrorKey"},
```
TODO: If this is an older SDK, maybe we should also leave this one as-is.

jelee53 replied:
Because this seems auto-generated, I'll delete this.

pvollan commented on Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
```diff
@@ -103,7 +103,7 @@ - (void)encodeWithCoder:(NSCoder *)coder
 
     if (!resourceError.failingURL().isEmpty()) {
-        [userInfo setValue:resourceError.failingURL().string().createNSString().get() forKey:@"NSErrorFailingURLStringKey"];
+        [userInfo setValue:resourceError.failingURL().string().createNSString().get() forKey:@"NSURLErrorFailingURLErrorKey"];
```
Should this set the value to an NSURL object instead?

pvollan commented on Source/WebCore/platform/network/cf/ResourceError.h:
```diff
@@ -71,7 +71,9 @@ class ResourceError : public ResourceErrorBase {
 
     WEBCORE_EXPORT ErrorRecoveryMethod errorRecoveryMethod() const;
 
+#if !HAVE_NSURL_ERROR_FAILING_URL_ERROR_KEY_ONLY
```
-> `HAVE(NSURL_ERROR_FAILING_URL_ERROR_KEY_ONLY)`

jelee53 replied:
From offline discussion with Per, renamed to `ENABLE_NSURL_ERROR_FAILING_URL_STRING_KEY_DEPRECATION` for clarity.

pvollan commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadInvalidURLRequest.mm:
```diff
@@ -176,7 +176,7 @@ - (void)encodeWithCoder:(NSCoder *)coder
         EXPECT_WK_STREQ(error.domain, @"WebKitErrorDomain");
         EXPECT_EQ(error.code, WebKitErrorCannotShowURL);
-        EXPECT_WK_STREQ(error.userInfo[NSErrorFailingURLStringKey], "https://%20.example.com/");
+        EXPECT_WK_STREQ([error.userInfo[NSURLErrorFailingURLErrorKey] absoluteString], "");
```
Do you need to check `HAVE(NSURL_ERROR_FAILING_URL_ERROR_KEY_ONLY)` here as well?

jelee53 replied:
From offline discussion with Per, we will not add it because `NSErrorFailingURLStringKey` is deprecated when this test is created,

pvollan commented on Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
```diff
@@ -227,7 +228,7 @@ - (void)encodeWithCoder:(NSCoder *)coder
 
     RetainPtr userInfo = [m_platformError userInfo];
-    if (auto *failingURLString = dynamic_objc_cast<NSString>([userInfo valueForKey:@"NSErrorFailingURLStringKey"]))
+    if (auto *failingURLString = dynamic_objc_cast<NSString>([userInfo valueForKey:@"NSURLErrorFailingURLStringErrorKey"]))
```
Should we be using the symbol here, and not the string?

szewai replied:
Yes, it's either `@"NSErrorFailingURLStringKey"` or `NSURLErrorFailingURLStringErrorKey`; `@"NSURLErrorFailingURLStringErrorKey"` is not a valid key.

pvollan commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1105,3 +1105,11 @@
+    || (PLATFORM(VISION) && __VISION_OS_VERSION_MIN_REQUIRED >= 20500) \
+    || (PLATFORM(WATCHOS) && __WATCH_OS_VERSION_MIN_REQUIRED >= 110500) \
+    || (PLATFORM(APPLETV) && __TV_OS_VERSION_MIN_REQUIRED >= 180500))
+#define ENABLE_NSURL_ERROR_FAILING_URL_STRING_KEY_DEPRECATION 1
```
On second thought, I wonder if it's better to use `HAVE_NSURL_ERROR_FAILING_URL_STRING_KEY` or `USE_NSURL_ERROR_FAILING_URL_STRING_KEY`. What do you think?

szewai commented on Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
```diff
@@ -103,9 +103,10 @@ - (void)encodeWithCoder:(NSCoder *)coder
+#if !ENABLE(NSURL_ERROR_FAILING_URL_STRING_KEY_DEPRECATION)
+        [userInfo setValue:resourceError.failingURL().string().createNSString().get() forKey:NSURLErrorFailingURLStringErrorKey];
+#endif
+        [userInfo setValue:resourceError.failingURL().createNSURL().get() forKey:NSURLErrorFailingURLErrorKey];
```
Why is `if (RetainPtr cocoaURL = resourceError.failingURL().createNSURL())` removed?

jelee53 replied:
I was thinking that even if the cocoaURL is null, NSURLErrorFailingURLErrorKey should be created regardless. 

But I will add it back (so that this key is only created if you can create a valid NSURL from the failing URL). As you were saying, developers will only rely on the error code then.

szewai commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -432,3 +432,11 @@
 #define USE_MODERN_AVCONTENTKEYSESSION_WITH_VTDECOMPRESSIONSESSION 1
 #endif
+
+#if ((PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED < 150500) \
```
Should this be `<=`?

jelee53 replied:
Oh that's a fair comment. Thanks!

It got deprecated in iOS 18.4 so I assumed the next version that wouldn't have it is 18.5. But you're right that if there's like 18.4.1 or something, it would get accidentally included.
https://developer.apple.com/documentation/foundation/nsurlerrorfailingurlstringerrorkey


I'll re-correct it to be <= "the actual version it got deprecated in".

szewai replied:
Oh sorry I meant to say `<= 150400`

pvollan commented on Source/WebKit/Shared/Cocoa/WebErrorsCocoa.mm:
```diff
@@ -37,10 +37,10 @@
+    RetainPtr<NSMutableDictionary> userInfo = [NSMutableDictionary dictionaryWithObjectsAndKeys:URL, NSURLErrorFailingURLErrorKey, nil];
+#if USE(NSURL_ERROR_FAILING_URL_STRING_KEY)
+    [userInfo setValue:[URL absoluteString] forKey:NSURLErrorFailingURLStringErrorKey];
+#endif
```
Perhaps you don't need to change this to a mutable dictionary?

pvollan replied:
You may be able to guard `[URL absoluteString], NSURLErrorFailingURLStringErrorKey,` with `#if USE(NSURL_ERROR_FAILING_URL_STRING_KEY)`

jelee53 replied:
Oh I didn't know you could break up a literal like that! Done.

pvollan commented on Source/WebKit/Shared/Cocoa/WebErrorsCocoa.mm:
```diff
@@ -39,7 +39,9 @@
     RetainPtr<NSDictionary> userInfo = [NSDictionary dictionaryWithObjectsAndKeys:
         URL, NSURLErrorFailingURLErrorKey,
+#if USE(NSURL_ERROR_FAILING_URL_STRING_KEY)
         [URL absoluteString], @"NSErrorFailingURLStringKey",
```
Should we use `NSURLErrorFailingURLStringErrorKey` here?

jelee53 replied:
Sorry! I wasn't very careful when I reverted that hunk. I will fix it!

pvollan replied:
NP! Thanks for fixing :)

jelee53 replied:
Fixed! Thanks for catching!

jelee53 committed:
Replace deprecated "NSErrorFailingURLStringKey" and "NSURLErrorFailingURLStringErrorKey" with "NSURLErrorFailingURLErrorKey"
https://bugs.webkit.org/show_bug.cgi?id=297139
rdar://158015996

Reviewed by Sihui Liu and Per Arne Vollan.

Remove usage of  "NSErrorFailingURLStringKey" and "NSURLErrorFailingURLStringErrorKey" in up-to-date platforms

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/network/cf/ResourceError.h:
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::createNSErrorFromResourceErrorBase):
(WebCore::ResourceError::hasMatchingFailingURLKeys const):
* Source/WebKit/Shared/Cocoa/WebErrorsCocoa.mm:
(WebKit::createNSError):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadInvalidURLRequest.mm:
(TestWebKitAPI::TEST(WebKit, LoadInvalidURLWithSpaceCharacter)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeBrowsing.mm:
(TEST(SafeBrowsing, WKWebViewGoBackIFrame)):
(TEST(SafeBrowsing, PostResponseIframe)):

Canonical link: https://commits.webkit.org/298727@main
